### PR TITLE
DC-279: Renew certificate

### PR DESCRIPTION
**Pull request by patrick-schweizer**
Jira: https://issues.opennms.org/browse/DC-279

### CODE REVIEW

File: plugin/src/main/java/org/opennms/plugins/cloud/rest/CloudConfigRestServiceImpl.java [hunk 1]
```diff
@@ -43,9 +43,16 @@ public CloudConfigRestServiceImpl(final ConfigurationManager cm) {
...
-        this.cm.initConfiguration(key);
-        this.cm.configure();
+        try {
+            String key = extractKey(keyJson);
+            this.cm.initConfiguration(key);
+            this.cm.configure();
+        } catch (Exception e) {
+            return Response
+                    .status(Response.Status.INTERNAL_SERVER_ERROR)
+                    .entity("{\"status\":\"" + cm.getStatus() + "\", \"message\":\"" + e.getMessage() + "}")
```
Comment (freddy61025) [inline plugin/src/main/java/org/opennms/plugins/cloud/rest/CloudConfigRestServiceImpl.java:53]: may be use String.format or JSONObject to make it easier to read.

**Review by freddy61025 — DISMISSED** [review]
Some minor problem only. 

BTW I want to ask about the README Configuration Properties part still valid? By my testing it will be overridden by the property in zip file.

File: plugin/src/test/java/org/opennms/plugins/cloud/config/HousekeeperTest.java [hunk 2]
```diff
@@ -46,13 +47,14 @@ public class HousekeeperTest {
     private Housekeeper hk;
 
     @Test
-    public void shouldRenewConfigForExpiredToken() throws InterruptedException {
+    public void shouldRenewConfigForExpiredToken() {
         ConfigurationManager cm = mock(ConfigurationManager.class);
         hk = new Housekeeper(cm, 1);
-        doReturn(Instant.now()
-                .plusSeconds(60*60)) // first time: token valid
+        doReturn(Instant.now().plusSeconds(60*60)) // first time: token valid
```
[the remark this answers is not in the document]
Comment (patrick-schweizer) [reply plugin/src/test/java/org/opennms/plugins/cloud/config/HousekeeperTest.java]: fixed

File: plugin/src/test/java/org/opennms/plugins/cloud/config/HousekeeperTest.java [hunk 3]
```diff
@@ -46,13 +47,14 @@ public class HousekeeperTest {
...
                 .doReturn(Instant.now()) // second time: token expired
                 .when(cm).getTokenExpiration();
+        doReturn(Instant.now().plusSeconds(60*60)).when(cm).getCertExpiration(); // cert always valid
```
[the remark this answers is not in the document]
Comment (patrick-schweizer) [reply plugin/src/test/java/org/opennms/plugins/cloud/config/HousekeeperTest.java]: fixed

File: plugin/src/main/java/org/opennms/plugins/cloud/config/Housekeeper.java [hunk 4]
```diff
@@ -64,10 +66,32 @@ public void destroy() {
...
+        if (expirationDate.minusSeconds(60L * 20L).isBefore(Instant.now())) {
             log.info("Triggering renewal of configuration, token will expire soon.");
             this.configurationManager.configure();
         }
     }
 
+    public void renewCerts() throws CertificateException {
+        final Instant expirationDate = configurationManager.getCertExpiration();
+        // renew 20 min before expiry
+        if(expirationDate.minusSeconds(60L * 20L).isBefore(Instant.now())) {
```
[the remark this answers is not in the document]
Comment (patrick-schweizer) [reply plugin/src/main/java/org/opennms/plugins/cloud/config/Housekeeper.java]: good point! 
I changed the intervals for cert and token

File: plugin/src/main/java/org/opennms/plugins/cloud/config/Housekeeper.java [hunk 5]
```diff
@@ -55,7 +56,8 @@ public Housekeeper(final ConfigurationManager configurationManager) {
 
 
     public void init() {
-        executor.scheduleAtFixedRate(this::renewToken, intervalInSeconds, intervalInSeconds, TimeUnit.SECONDS);
+        executor.scheduleAtFixedRate(() -> wrap(this::renewToken), intervalInSeconds, intervalInSeconds, TimeUnit.SECONDS);
+        executor.scheduleAtFixedRate(() -> wrap(this::renewCerts), intervalInSeconds, intervalInSeconds, TimeUnit.SECONDS);
```
[the remark this answers is not in the document]
Comment (patrick-schweizer) [reply plugin/src/main/java/org/opennms/plugins/cloud/config/Housekeeper.java]: I changed not to check at different intervals

File: plugin/src/main/java/org/opennms/plugins/cloud/rest/CloudConfigRestServiceImpl.java (hunk 1 above)
Comment (patrick-schweizer) [reply plugin/src/main/java/org/opennms/plugins/cloud/rest/CloudConfigRestServiceImpl.java]: good point. I replaced it with String.format (didn't want to get into the JSon dependency hell just yet :-D

**Review by freddy61025 — APPROVED** [review]
LGTM

**Comment by freddy61025** [comment]
BTW for the coverage is it possible to set some exclude in sonacloud.

**Comment by patrick-schweizer** [comment]
> BTW for the coverage is it possible to set some exclude in sonacloud.

Good to know!

**Pull request merged**